include/diy/mpi/collectives.hpp

```diff
@@ -75,15 +76,16 @@ namespace mpi
     {
 #ifndef DIY_NO_MPI
       std::vector<int>  counts(comm.size());
-      Collectives<int,void*>::gather(comm, (int) in.size(), counts, root);
+      int               elem_size = Datatype::count(in[0]);     // size of 1 vector element in units of mpi datatype
+      Collectives<int,void*>::gather(comm, (int)(elem_size * in.size()), counts, root);
```

Comment:
I don't think you should be multiplying by `elem_size` here. You just want to gather vector sizes. Then you don't need all the awkward division below.

Reply:
Ah, no. I see the point. You need `offsets` and `counts` in terms of the datatype.